sql/vector_mhnsw.cc

```diff
@@ -270,185 +174,261 @@ static bool select_neighbours(TABLE *source, TABLE *graph,
-    return 0;
+    if (node.get_ref_len() != ref_length)
+    {
+      // this should never happen
+      return true;
+    }
+    memcpy(neighbor_array->neigh_ref + i * ref_length, node.get_ref(),
+           node.get_ref_len());
+    i++;
   }
 
-  /*
-  if ((err= graph->file->ha_index_next_same(graph->record[0], key,
-                                            graph->key_info->key_length)))
-    return err == HA_ERR_END_OF_FILE ? 0 : -1;
-  */
-  return 1;
-}
-
-static inline bool store_link(TABLE *graph,
-                              size_t layer_number,
-                              const FVectorRef &src,
-                              const FVectorRef &dst)
-{
-  //TODO(cvicentiu) this causes 2 uneeded stores because field 0 and 1 almost
-  //never change.
-  //Or we could iterate and store them all src->dest first and then all dest->src
   graph->field[0]->store(layer_number);
   graph->field[1]->store_binary(
-    reinterpret_cast<const char *>(src.get_ref()), src.get_ref_len());
-  graph->field[2]->store_binary(
-    reinterpret_cast<const char *>(dst.get_ref()), dst.get_ref_len());
-  return graph->file->ha_write_row(graph->record[0]);
-}
-
-
-static bool write_connection_to_graph(TABLE *graph,
-                                      uchar *key,
-                                      size_t layer_number,
-                                      const FVectorRef &src,
-                                      const FVectorRef &dst)
-{
-  // TODO(cvicentiu) if store_link fails on the second write, we can end up
-  // with a corrupt index. Handle this and write undo logic.
-  int con_exists= connection_exists(graph, key, layer_number, src, dst);
-  if (con_exists)
-    return con_exists < 0;
-
-  return store_link(graph, layer_number, src, dst) ||
-         store_link(graph, layer_number, dst, src);
-}
+    reinterpret_cast<const char *>(source_node.get_ref()),
+    source_node.get_ref_len());
+  graph->field[2]->set_null();
 
-static bool insert_node_connections_on_layer(TABLE *graph,
-                                             size_t layer_number,
-                                             const FVector &source_node,
-                                             const List<FVector> &dest_nodes)
-{
   uchar *key= (uchar*)alloca(graph->key_info->key_length);
+  key_copy(key, graph->record[0], graph->key_info, graph->key_info->key_length);
 
-  if (dest_nodes.elements == 0)
+  int err= graph->file->ha_index_read_map(graph->record[1], key,
+                                           HA_WHOLE_KEY,
+                                           HA_READ_KEY_EXACT);
+
+  // no record
+  if (err)
   {
-    return store_link(graph, layer_number, source_node, source_node);
+    graph->field[2]->store_binary(
+      reinterpret_cast<const char *>(neighbor_array), totalSize);
+    graph->file->ha_write_row(graph->record[0]);
+    return false;
   }
 
-  for (const auto &node: dest_nodes)
+  // record need update
+  if (cmp_record(graph, record[1]))
```

Comment:
This compares full record, even the "uninitialized bytes" of the varbinary column. The low recall I was experiencing with your branch got fixed when removing this line.

Reply:
Nice catch! I initially had following two lines immediately after reading record to record[1] with `ha_index_read_map`. Later I was trying to optimize it to avoid storing field2 in some scenarios. But apparently the logic of storing after comparison is incorrect.

> ```
>     graph->field[2]->store_binary(
>       reinterpret_cast<const char *>(neighbor_array), totalSize);

Reply:
Another thing regarding the recall is that I tried various parameter values last week. I tested your original implementation, your implementation with my changes in this [PR](https://github.com/MariaDB/server/pull/3156), and the implementation in the current PR. However, I always got lower recall rate compared to PGVector using the same parameter sets. There could be something overlooked by us or done differently by PGVector.